### PR TITLE
[RPA-1668] Add sending decrypted credentials

### PR DESCRIPTION
**Pull request by marcin-depcik**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-scheduler/src/database/process/process.service.ts [hunk 1]
```diff
@@ -4,7 +4,17 @@ import { Repository } from 'typeorm';
...
+const relations = [
+    'createdBy',
+    'system',
+    'botCollection',
+    'schedules',
+    'editor',
+    'notifications.user.authorities',
+    'credentials.attributes',
+    'credentials.attributes.secret',
+    'credentials.template',
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/database/process/process.service.ts]: to rethink - maybe we want to use credentials.attributes.secret only where we need it - not everytime we GET process

File: runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts [hunk 2]
```diff
@@ -148,6 +149,9 @@ export class RuntimeService implements OnApplicationBootstrap, OnModuleDestroy {
             ...(request.userId && { user: { id: request.userId } }),
             ...(request.callbackUrl && { callbackUrl: request.callbackUrl }),
         };
+
+        delete processInstance?.decryptedCredentials;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts]: function programming

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts]: change to const { ...credentials, ...processInstanceWithoutCredentials } = processInstnace

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/runbotics-scheduler/src/database/process/process.service.ts (hunk 1 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/database/process/process.service.ts]: we want to keep both encrypted and decrypted credentials on sheduler only, and never send them to UI or keep em on the endpoint

Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/database/process/process.service.ts]: only to send decrypted to bot when it's supposed to run new process

**Comment by star-in-the-jar** [comment]
todo: 
- add credentialId in the list
- check if credential is primary based on the order in db
- make credentials list required

File: runbotics/runbotics-scheduler/package.json [hunk 3]
```diff
@@ -67,7 +67,8 @@
         "cross-fetch": "^3.1.5",
         "@nestjs-modules/mailer": "1.8.1",
         "zod": "~3.23.8",
-        "unplugin-swc": "~1.4.5"
+        "unplugin-swc": "~1.4.5",
+        "bpmn-moddle": "^9.0.1"
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/package.json:71]: noice 👌

File: runbotics/runbotics-scheduler/src/database/process/process.service.ts [hunk 4]
```diff
@@ -15,13 +23,20 @@ interface PartialUpdateProcess extends IProcess {
...
+    ) {}
 
     findById(id: number): Promise<IProcess> {
         return this.processRepository.findOne({ where: { id }, relations });
     }
 
+    findByIdWithSecrets(id: number) {
+        return this.processRepository.findOne({
+            where: { id },
+            relations: [...relations, 'credentials.attributes.secret'],
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/database/process/process.service.ts]: great

File: runbotics/runbotics-scheduler/src/queue/process/process-scheduler.service.ts [hunk 5]
```diff
@@ -1,41 +1,176 @@
...
+                    await this.getDecryptedCredentials(
+                        subprocessId,
+                        subprocessDefinition
+                    );
+
+                return subprocessDecryptedCredentials;
+            })
+        );
+
+        return subprocessDecryptedCredentials.flatMap((c) => c);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/queue/process/process-scheduler.service.ts]: just .flat() is enough here

File: runbotics/runbotics-scheduler/src/queue/process/process-scheduler.service.ts [hunk 6]
```diff
@@ -1,41 +1,176 @@
...
+    }
+
+    private filterDuplicatedCredentials(
+        decryptedCredentials: DecryptedCredential[]
+    ): DecryptedCredential[] {
+        return decryptedCredentials.filter(
+            (credential, index, array) =>
+                index === array.findIndex((c) => credential.id === c.id)
+        );
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/queue/process/process-scheduler.service.ts]: O(n^2)
if you create Set with id as a key and all credential data as a value you have O(n)

Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/queue/process/process-scheduler.service.ts]: or if you want to keep the array, you can use set.has(credential.id) as filter condition

File: runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import BpmnModdle, { BaseElement, Definitions } from 'bpmn-moddle';
+import { GeneralAction, IProcess } from 'runbotics-common';
+
+interface CustomExtensionElement {
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts]: "Custom" doesn't feel like describing enough, but It's hard to suggest sth else

Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts]: maybe "Parsed"

File: runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    element !== null &&
+    element !== undefined &&
+    typeof element === 'object' &&
+    '$attrs' in element &&
+    element.$attrs !== null &&
+    element.$attrs !== undefined &&
+    typeof element.$attrs === 'object' &&
+    'camunda:actionId' in element.$attrs &&
+    typeof element.$attrs['camunda:actionId'] === 'string' &&
+    element.$attrs['camunda:actionId'] === GeneralAction.START_PROCESS;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts]: whoa
consider breaking this down into smaller parts like

const isObject =  
    elementId.startsWith('Activity') &&
    element !== null &&
    element !== undefined &&
    typeof element === 'object'

const hasAttributes=
'$attrs' in element &&
    element.$attrs !== null &&
    element.$attrs !== undefined &&
    typeof element.$attrs === 'object'

const hasStartEventActionId = 
'camunda:actionId' in element.$attrs &&
    typeof element.$attrs['camunda:actionId'] === 'string' &&
    element.$attrs['camunda:actionId'] === GeneralAction.START_PROCESS;

const isActionStartProcess = 
isObject &&
hasAttributes &&
hasStartEventActionId

File: runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                        ({ name }) => name === 'processId'
+                    );
+                if (!startProcessActionElement) return processIds;
+
+                const processId = startProcessActionElement.$body.match(
+                    startProcessActionBodyPattern
+                );
+                if (!processId) return processIds;
+
+                processIds.push(Number(processId[1]));
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/utils/bpmn/findSubprocess.ts]: why '1'? could be extracted to named const for readability

File: runbotics/runbotics-sdk/src/handler/DesktopRunRequest.ts [hunk 10]
```diff
@@ -8,6 +8,16 @@ interface CommonDesktopRunRequest {
...
+    credentials: {
+        id: string;
+        name: string;
+        template: string;
+        attributes: {
+            id: string;
+            name: string;
+            value: string;
+        }[];
+    }[];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-sdk/src/handler/DesktopRunRequest.ts:20]: If i remember correctly @Aeri4a created credential interface in rb-common - credential should be defined in one place, so can be reused, and any changes would appear in the whole project. If you're merging before Ania, leave @TODO comment

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/runbotics-sdk/src/handler/DesktopRunRequest.ts (hunk 10 above)
Comment (marcin-depcik) [reply runbotics/runbotics-sdk/src/handler/DesktopRunRequest.ts:20]: In `runbotics-sdk` we cannot use dependency  `runbotics-common` which is not hosted package

**Pull request merged**